### PR TITLE
add 'predeclared' as a non-default linter

### PR DESCRIPTION
**Pull request by nishanths**
The predeclared command checks for code that shadows Go's predeclared
identifiers.

godoc: https://godoc.org/github.com/nishanths/predeclared
source: https://github.com/nishanths/predeclared

The list of predeclared identifiers is in [the spec](https://golang.org/ref/spec#Predeclared_identifiers).

See related issue #417 for more details.

### CODE REVIEW

**Comment by nishanths** [comment]
TODO: fix non availability of package `go/doc` in go1.7?

File: linters.go [hunk 1]
```diff
@@ -340,6 +340,13 @@ var defaultLinters = map[string]LinterConfig{
 		InstallFrom:       "github.com/alexkohler/nakedret",
 		PartitionStrategy: partitionPathsAsDirectories,
 	},
+	"predeclared": {
+		Command:           `predeclared`,
+		Pattern:           `PATH:LINE:COL:MESSAGE`,
+		InstallFrom:       "github.com/nishanths/predeclared",
+		PartitionStrategy: partitionPathsAsDirectories,
+		IsFast:            true,
```
Comment (nishanths) [inline linters.go]: Using `true` based on the fact that the speed is comparable to ineffassign, which is marked as fast:

```sh
$ time ineffassign /usr/local/go/src/net/
/usr/local/go/src/net/dial_test.go:338:6: ineffectual assignment to err
/usr/local/go/src/net/fd_plan9.go:73:4: ineffectual assignment to err
/usr/local/go/src/net/http/fs.go:604:5: ineffectual assignment to name
/usr/local/go/src/net/http/request_test.go:147:13: ineffectual assignment to err
/usr/local/go/src/net/http/serve_test.go:4452:13: ineffectual assignment to err
/usr/local/go/src/net/http/transport_test.go:716:8: ineffectual assignment to err
/usr/local/go/src/net/http/transport_test.go:2452:3: ineffectual assignment to slurp
/usr/local/go/src/net/interface_plan9.go:155:3: ineffectual assignment to line
/usr/local/go/src/net/interface_plan9.go:155:9: ineffectual assignment to ok
/usr/local/go/src/net/udpsock_plan9.go:80:16: ineffectual assignment to i

real	0m0.401s
user	0m0.378s
sys	0m0.027s
```

```sh
$ time predeclared /usr/local/go/src/net/
/usr/local/go/src/net/file_plan9.go:34:2: variable close has same name as predeclared identifier
/usr/local/go/src/net/http/clientserver_test.go:1025:12: variable close has same name as predeclared identifier
/usr/local/go/src/net/http/server.go:1927:30: param error has same name as predeclared identifier
/usr/local/go/src/net/lookup.go:76:56: named return error has same name as predeclared identifier
/usr/local/go/src/net/mac.go:83:1: label error has same name as predeclared identifier
/usr/local/go/src/net/rpc/server.go:207:2: variable rune has same name as predeclared identifier

real	0m0.292s
user	0m0.229s
sys	0m0.033s
```

**Comment by nishanths** [comment]
fixed tests. this should be ready to review.

File: linters.go [hunk 2]
```diff
@@ -340,6 +340,13 @@ var defaultLinters = map[string]LinterConfig{
 		InstallFrom:       "github.com/alexkohler/nakedret",
 		PartitionStrategy: partitionPathsAsDirectories,
 	},
+	"predeclared": {
```
Comment (nishanths) [inline linters.go]: open to setting `defaultEnabled: true` if that is preferred

**Comment by alecthomas** [comment]
Thanks for doing this, however please take a look at #380. We are trying to shift towards API-based integration with any new linters.

Because of the Christmas break we haven't had time to implement the API on the gometalinter side, but hopefully we can get to that soon and then integrate `predeclared` via that mechanism.

**Comment by alecthomas** [comment]
I forgot to mention: this is actually a really great linter to have, as this is a pretty nasty potential source of bugs.

**Comment by nishanths** [comment]
SGTM!

**Pull request closed without merging**